### ISSUE TITLE
npm install from github Enterprise Private Mode

### ISSUE THREAD

**Issue by rhettl** [npm/npm#9408]
Hello, 

I have been trying to find other issue with this and even went to reading many closed issues, mine is similar to #8435, #8345, and I think #8031. 

We just setup a Github Enterprise server and by Github's mandate it has to be in "Private mode" meaning everyone has to log in to see anything marked public. 

I added a repo and want to share it with others, but `npm install  git+ssh://github.oit.uci.edu:education/node-ldap.git` doesn't work. here is a [gist of the log](https://gist.github.com/rhettl/0e32293e274bafb5d1cb).

I have tried changing versions to 2.1.x, latest, and a few others. I have updated node to latest and nothing. Bower works fine and so does git.

So far the only way to make it work is this:
```sh
$ GIT_SSH_COMMAND='ssh -i ~/.ssh/id_rsa.pub' npm install --save git+ssh://git@github.oit.uci.edu:education/node-ldap.git
```

Thanks for any help you can give,

**Comment by robmcl4** [comment]
Have you registered your ssh public key on github enterprise? If you haven't, or otherwise don't have read access to the repository, then you would get this message I presume.

**Comment by rhettl** [comment]
Sorry, I could have sworn I said that. Yes, my key is in Github Enterprise, on my computer, and associated with that repo and with full administrative privileges. The same key is the one used in the work around example code 

> ```js
$ GIT_SSH_COMMAND='ssh -i ~/.ssh/id_rsa.pub' npm install --save git+ssh://git@github.oit.uci.edu:education/node-ldap.git
```

**Comment by othiym23** [comment]
We've made a bunch of tweaks to our Git support since `npm@2.7.1` -- what happens if you [update to the latest version of npm](https://github.com/npm/npm/wiki/Troubleshooting#try-the-latest-stable-version-of-npm)?

**Comment by rhettl** [comment]
> ```
$ sudo npm install -g npm@latest
/usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
npm WARN unmet dependency /usr/local/lib/node_modules/generator/node_modules/replace requires colors@'0.5.x' but will load
npm WARN unmet dependency /usr/local/lib/node_modules/generator/node_modules/colors,
npm WARN unmet dependency which is version 0.6.2
npm@2.14.1 /usr/local/lib/node_modules/npm
```

The update to `npm@latest` worked and now allows the download. Oddly, I did that exact update 3 days ago before posting here. I tried `npm@latest` with both `node@0.10.x` and `node@0.12.x` with the same result as the original post. 

I may need to make a new post for this, but I am still getting some errors, I think they are unrelated to Git and Private Mode, but in case these are related, here is the new issue.

On installing the overall package `npm install --save git+ssh://git@github.oit.uci.edu:education/node-ldap.git` I get a large amount of errors after the `node-gyp rebuild` command from node-gyp. If I run `npm install dtrace-provider` it doesn't run a rebuild and instead runs `node scripts/install.js` and runs with no errors. [Here is a gist](https://gist.github.com/rhettl/46599bcf1c773180a8e5) with my package.json and two commands with verbose flag.

Thanks @othiym23 and @robmcl4, I appreciate your help so far.

**Comment by othiym23** [comment]
I bet you're using io.js! There are floating patches applied to the version of `node-gyp` included with the version of npm packaged with iojs necessary to make it possible for `node-gyp` to download the iojs headers it needs to run native builds . The easiest way to work around this (for now) is to `npm install -g pangyp; npm config set node-gyp pangyp` and retry your install (and remember you did that, because when Node 4.0.0 and Node 5.0.0 come out (soon!) you'll want to run `npm config del node-gyp` to remove it again).

**Comment by rhettl** [comment]
I shouldn't be, I uninstalled io.js months ago due to issues with
htmlWebkitToPdf. Also I just updated to node 0.12.latest. If it is
connected to io.js, I'm not die how. How might I check of npm is utilizing
a versing of io.js issued of node?

**Comment by rhettl** [comment]
Okay, figured out how to check, 

```
$ node-gyp ?

  Usage: node-gyp <command> [options]

  where <command> is one of:
    - build - Invokes `make` and builds the module
    - clean - Removes any generated build files and the "out" dir
    - configure - Generates a Makefile for the current module
    - rebuild - Runs "clean", "configure" and "build" all at once
    - install - Install node development files for the specified node version.
    - list - Prints a listing of the currently installed node development files
    - remove - Removes the node development files for the specified version

node-gyp@2.0.2  /usr/local/lib/node_modules/node-gyp
node@0.12.7
```

So the help screen says I am in fact using `node@0.12.7`

**Comment by ehsalazar** [comment]
It sounds like updating did the trick. I'll close this for now as we’re trying to clean up some solved issues. We’ll be happy to reopen if necessary.

**Comment by mcpDESIGNS** [comment]
@rhettl hey, sorry to bother you! I was about to try to create an npm package myself for our enterprise as well. No one outside, who can't access the enterprise irk can see the package, correct?

**Comment by rhettl** [comment]
So to update what happened after the closing of this ticket...

The issues I described here were fixed, mostly. I ran into another roadblock with my company which was the VPN/Firewall. Since my company has a very restrictive firewall around our private GH server, despite having it LDAP password protected (a redundancy I lobby against), people/machines must be inside our network or connected through VPN before they can `git clone` making `npm install uname/repo` or other usages mostly impossible unless you want to connect via VPN *every* time you update. 

After a year of seeking out my company's copyright decision maker, I had a meeting with him to approve a blanket permission for public releases (given some prerequisites). I made a public github org and put my code there.

I hope that helps a little. At the moment I think there is too little information to tell you much. Offhand it sounds like my issue with the firewall. 

If you want, I am happy to help you in a less public place. I updated my github account with my new website which has an email form.
